File: src/domain/onboarding/service.ts
import type { PoolClient } from 'pg';
import { withTx } from '../../db/pool';
import { ApiError } from '../../errors';
import { writeAudit } from '../../audit/writeAudit';
import { documentStore } from '../../storage';
import type {
  ApprovedAccountSummary,
  BusinessDetails,
  IndustrySector,
  OnboardingApplication,
  UploadedDocument,
} from '../../contract';
import type { RequestSession } from '../../http/auth';
import type { BusinessDetailsInput, PrincipalInput } from './schemas';
import * as repo from './repo';
import { kybProvider } from './kyb';

const ALLOWED_MIME_TYPES = new Set(['application/pdf', 'image/jpeg', 'image/png']);
const MAX_FILE_SIZE_BYTES = 10 * 1024 * 1024;

// Lifted from Kimana_frontend/src/api/mock/onboardingApi.ts.
const INDUSTRY_SEGMENT_LABEL: Record<IndustrySector, string> = {
  agriculture_agro_export: 'Agro Exporter',
  textiles_apparel: 'Textiles Exporter',
  solid_minerals: 'Solid Minerals Exporter',
  manufacturing: 'Manufacturing Exporter',
  oil_gas_services: 'Oil & Gas Services',
  technology: 'Technology Exporter',
  trading_commodities: 'Commodities Trader',
  other: 'Trading Business',
};

function generateAccountId(legalName: string): string {
  const initials = legalName
    .split(/\s+/)
    .filter(Boolean)
    .slice(0, 3)
    .map((word) => word[0]?.toUpperCase() ?? '')
    .join('');
  const serial = Math.floor(Math.random() * 90000 + 10000);
  return `${initials || 'KMA'}-${serial}`;
}

/**
 * Loads the caller's application and, if `applicationId` was supplied, checks
 * it matches — mirroring the mock's `NOT_FOUND` on a wrong id.
 */
async function loadOwnedApplication(
  session: RequestSession,
  applicationId?: string,
): Promise<OnboardingApplication> {
  const application = await repo.findApplicationByCustomer(session.customerId);
  if (!application) throw ApiError.notFound('No onboarding application for this customer.');
  if (applicationId && applicationId !== application.id) {
    throw ApiError.notFound('Application not found.');
  }
  return application;
}

export async function getApplication(session: RequestSession): Promise<OnboardingApplication> {
  return loadOwnedApplication(session);
}

export async function saveBusinessDetails(
  session: RequestSession,
  business: BusinessDetailsInput,
  applicationId?: string,
): Promise<OnboardingApplication> {
  const application = await loadOwnedApplication(session, applicationId);

  return withTx(async (client) => {
    await repo.saveBusiness(client, application.id, business as BusinessDetails);
    await writeAudit(client, {
      actorId: session.userId,
      actorRole: session.role,
      action: 'onboarding.business_saved',
      entityType: 'onboarding_application',
      entityId: application.id,
      before: application.business,
      after: business,
    });
    return repo.reloadApplication(client, application.id);
  });
}

export async function savePrincipals(
  session: RequestSession,
  principals: readonly PrincipalInput[],
  applicationId?: string,
): Promise<OnboardingApplication> {
  const application = await loadOwnedApplication(session, applicationId);

  return withTx(async (client) => {
    await repo.replacePrincipals(client, application.id, principals);
    await writeAudit(client, {
      actorId: session.userId,
      actorRole: session.role,
      action: 'onboarding.principals_saved',
      entityType: 'onboarding_application',
      entityId: application.id,
      before: application.principals,
      after: principals,
    });
    return repo.reloadApplication(client, application.id);
  });
}

export interface DocumentUpload {
  type: UploadedDocument['type'];
  fileName: string;
  mimeType: string;
  buffer: Buffer;
}

function assertFile(mimeType: string, sizeBytes: number): void {
  if (!ALLOWED_MIME_TYPES.has(mimeType)) {
    throw ApiError.validation('That file type isn’t supported. Upload a PDF, JPG, or PNG.');
  }
  if (sizeBytes > MAX_FILE_SIZE_BYTES) {
    throw ApiError.validation('That file is larger than 10 MB. Compress it or choose a smaller copy.');
  }
}

export async function uploadDocument(
  session: RequestSession,
  upload: DocumentUpload,
  applicationId?: string,
): Promise<UploadedDocument> {
  const application = await loadOwnedApplication(session, applicationId);
  assertFile(upload.mimeType, upload.buffer.byteLength);

  const storageKey = `onboarding/${application.id}/${upload.type}/${Date.now()}-${upload.fileName}`;
  await documentStore.put(storageKey, upload.buffer, upload.mimeType);

  const previous = application.documents.find((d) => d.type === upload.type);

  const document = await withTx(async (client) => {
    const saved = await repo.upsertUploadedDocument(client, application.id, {
      type: upload.type,
      fileName: upload.fileName,
      mimeType: upload.mimeType,
      sizeBytes: upload.buffer.byteLength,
      storageKey,
    });
    await writeAudit(client, {
      actorId: session.userId,
      actorRole: session.role,
      action: previous ? 'onboarding.document_replaced' : 'onboarding.document_uploaded',
      entityType: 'onboarding_document',
      entityId: saved.id,
      before: previous ?? null,
      after: saved,
    });
    return saved;
  });

  return document;
}

export async function retryDocumentUpload(
  session: RequestSession,
  documentId: string,
  applicationId?: string,
): Promise<UploadedDocument> {
  const application = await loadOwnedApplication(session, applicationId);
  const found = await repo.findDocument(documentId);
  if (!found || found.applicationId !== application.id) {
    throw ApiError.notFound('Document not found.');
  }

  // Slice: bytes are already stored on the original upload, so retry just
  // clears the failed state. A real retry re-accepts the file.
  return withTx(async (client) => {
    const updated = await repo.markDocumentUploaded(client, documentId);
    await writeAudit(client, {
      actorId: session.userId,
      actorRole: session.role,
      action: 'onboarding.document_retried',
      entityType: 'onboarding_document',
      entityId: documentId,
      before: found.document,
      after: updated,
    });
    return updated;
  });
}

export async function removeDocument(
  session: RequestSession,
  documentId: string,
  applicationId?: string,
): Promise<void> {
  const application = await loadOwnedApplication(session, applicationId);
  const found = await repo.findDocument(documentId);
  if (!found || found.applicationId !== application.id) {
    throw ApiError.notFound('Document not found.');
  }

  await withTx(async (client) => {
    await repo.deleteDocument(client, documentId);
    await writeAudit(client, {
      actorId: session.userId,
      actorRole: session.role,
      action: 'onboarding.document_removed',
      entityType: 'onboarding_document',
      entityId: documentId,
      before: found.document,
      after: null,
    });
  });

  if (found.storageKey) {
    await documentStore.delete(found.storageKey).catch(() => undefined);
  }
}

/**
 * Slice: synchronous KYB that always approves. Walks the real status chain
 * (draft -> submitted -> in_review -> approved), writing an audit row per
 * transition, runs the KYB provider, then lands on `approved` (with
 * approvedSummary) or `rejected` (with rejectionReasons). The stub provider
 * always approves unless the data trips a documented trigger — see
 * kyb/stubProvider.ts.
 *
 * The HTTP request stays open until a terminal status is reached, matching the
 * contract ("resolves once checks complete with the final status") and the
 * frontend, which does not poll. A real provider that runs for minutes would
 * want this split into submit + webhook/poll — see docs/backend-plan.md §03.
 */
async function auditStateChange(
  client: PoolClient,
  session: RequestSession,
  applicationId: string,
  from: string,
  to: string,
): Promise<void> {
  await writeAudit(client, {
    actorId: session.userId,
    actorRole: session.role,
    action: 'onboarding.state_change',
    entityType: 'onboarding_application',
    entityId: applicationId,
    before: { status: from },
    after: { status: to },
  });
}

export async function submit(
  session: RequestSession,
  applicationId?: string,
): Promise<OnboardingApplication> {
  const application = await loadOwnedApplication(session, applicationId);

  if (!application.business) {
    throw ApiError.validation('Add your business details before submitting.');
  }
  if (application.status === 'submitted' || application.status === 'in_review') {
    throw ApiError.conflict('This application is already being reviewed.');
  }

  // draft -> submitted -> in_review, committed before the checks run so a
  // concurrent getApplication sees in_review.
  await withTx(async (client) => {
    await repo.patchStatus(client, application.id, { status: 'submitted', submittedAt: true });
    await auditStateChange(client, session, application.id, application.status, 'submitted');
    await repo.patchStatus(client, application.id, { status: 'in_review' });
    await auditStateChange(client, session, application.id, 'submitted', 'in_review');
  });

  const outcome = await kybProvider.runChecks(application);

  return withTx(async (client) => {
    await repo.replaceKybChecks(client, application.id, outcome.checks);

    if (outcome.approved) {
      const business = application.business!;
      const summary: ApprovedAccountSummary = {
        accountId: generateAccountId(business.legalName),
        riskRatingLabel: 'Medium-Low',
        segment: INDUSTRY_SEGMENT_LABEL[business.industry] ?? 'Trading Business',
        corridor: 'NGN → USD / EUR',
        monthlyLimit: { amountMinor: 100_000_00, currency: 'USD' },
      };
      await repo.patchStatus(client, application.id, {
        status: 'approved',
        reviewedAt: true,
        approvedSummary: summary,
      });
      await auditStateChange(client, session, application.id, 'in_review', 'approved');
    } else {
      await repo.patchStatus(client, application.id, {
        status: 'rejected',
        reviewedAt: true,
        rejectionReasons: outcome.rejectionReasons,
      });
      await auditStateChange(client, session, application.id, 'in_review', 'rejected');
    }

    return repo.reloadApplication(client, application.id);
  });
}

File: src/domain/onboarding/service.rs
use super::kyb;
use super::repo::{self, DocInput, StatusPatch};
use super::schema;
use crate::audit::{write_audit, AuditEntry};
use crate::contract::common::{CurrencyCode, Money};
use crate::contract::onboarding::{
    ApprovedAccountSummary, BusinessDetails, DirectorOrBeneficialOwner, OnboardingApplication,
    OnboardingDocumentType, UploadedDocument,
};
use crate::error::{ApiError, ApiResult};
use crate::http::Session;
use crate::state::AppState;
use crate::storage;
use crate::util::generate_account_id;
use chrono::Utc;
use serde_json::json;
use uuid::Uuid;

const ALLOWED_MIME: [&str; 3] = ["application/pdf", "image/jpeg", "image/png"];
const MAX_FILE_BYTES: i64 = 10 * 1024 * 1024;

async fn load_owned(
    state: &AppState,
    session: &Session,
    application_id: Option<&str>,
) -> ApiResult<OnboardingApplication> {
    let app = repo::find_by_customer(&state.pool, session.customer_id)
        .await?
        .ok_or_else(|| ApiError::not_found("No onboarding application for this customer."))?;
    if let Some(id) = application_id {
        if id != app.id {
            return Err(ApiError::not_found("Application not found."));
        }
    }
    Ok(app)
}

pub async fn get_application(
    state: &AppState,
    session: &Session,
) -> ApiResult<OnboardingApplication> {
    load_owned(state, session, None).await
}

pub async fn save_business_details(
    state: &AppState,
    session: &Session,
    business: BusinessDetails,
    application_id: Option<String>,
) -> ApiResult<OnboardingApplication> {
    schema::validate_business(&business)?;
    let app = load_owned(state, session, application_id.as_deref()).await?;
    let app_uuid = Uuid::parse_str(&app.id).unwrap();

    let mut tx = state.pool.begin().await?;
    repo::save_business(&mut tx, app_uuid, &business).await?;
    write_audit(
        &mut tx,
        AuditEntry {
            actor_id: Some(session.user_id),
            actor_role: Some(session.role_str()),
            action: "onboarding.business_saved",
            entity_type: "onboarding_application",
            entity_id: app.id.clone(),
            before: app
                .business
                .as_ref()
                .map(|b| serde_json::to_value(b).unwrap()),
            after: Some(serde_json::to_value(&business)?),
        },
    )
    .await?;
    tx.commit().await?;

    load_owned(state, session, None).await
}

pub async fn save_principals(
    state: &AppState,
    session: &Session,
    principals: Vec<DirectorOrBeneficialOwner>,
    application_id: Option<String>,
) -> ApiResult<OnboardingApplication> {
    schema::validate_principals(&principals)?;
    let app = load_owned(state, session, application_id.as_deref()).await?;
    let app_uuid = Uuid::parse_str(&app.id).unwrap();

    let mut tx = state.pool.begin().await?;
    repo::replace_principals(&mut tx, app_uuid, &principals).await?;
    write_audit(
        &mut tx,
        AuditEntry {
            actor_id: Some(session.user_id),
            actor_role: Some(session.role_str()),
            action: "onboarding.principals_saved",
            entity_type: "onboarding_application",
            entity_id: app.id.clone(),
            before: Some(serde_json::to_value(&app.principals)?),
            after: Some(serde_json::to_value(&principals)?),
        },
    )
    .await?;
    tx.commit().await?;

    load_owned(state, session, None).await
}

pub struct DocumentUpload {
    pub doc_type: OnboardingDocumentType,
    pub file_name: String,
    pub mime_type: String,
    pub bytes: Vec<u8>,
}

pub async fn upload_document(
    state: &AppState,
    session: &Session,
    upload: DocumentUpload,
    application_id: Option<String>,
) -> ApiResult<UploadedDocument> {
    let app = load_owned(state, session, application_id.as_deref()).await?;
    if !ALLOWED_MIME.contains(&upload.mime_type.as_str()) {
        return Err(ApiError::validation(
            "That file type isn't supported. Upload a PDF, JPG, or PNG.",
        ));
    }
    if upload.bytes.len() as i64 > MAX_FILE_BYTES {
        return Err(ApiError::validation(
            "That file is larger than 10 MB. Compress it or choose a smaller copy.",
        ));
    }
    let app_uuid = Uuid::parse_str(&app.id).unwrap();
    let previous = app
        .documents
        .iter()
        .find(|d| d.r#type == upload.doc_type.as_str())
        .cloned();

    let storage_key = format!(
        "onboarding/{}/{}/{}-{}",
        app.id,
        upload.doc_type.as_str(),
        Utc::now().timestamp_millis(),
        upload.file_name
    );
    storage::put(&state.config.storage_dir, &storage_key, &upload.bytes).await?;

    let mut tx = state.pool.begin().await?;
    let saved = repo::upsert_document(
        &mut tx,
        app_uuid,
        DocInput {
            doc_type: upload.doc_type.as_str(),
            file_name: &upload.file_name,
            mime_type: &upload.mime_type,
            size_bytes: upload.bytes.len() as i64,
            storage_key: &storage_key,
        },
    )
    .await?;
    write_audit(
        &mut tx,
        AuditEntry {
            actor_id: Some(session.user_id),
            actor_role: Some(session.role_str()),
            action: if previous.is_some() {
                "onboarding.document_replaced"
            } else {
                "onboarding.document_uploaded"
            },
            entity_type: "onboarding_document",
            entity_id: saved.id.clone(),
            before: previous.as_ref().map(|d| serde_json::to_value(d).unwrap()),
            after: Some(serde_json::to_value(&saved)?),
        },
    )
    .await?;
    tx.commit().await?;
    Ok(saved)
}

pub async fn retry_document_upload(
    state: &AppState,
    session: &Session,
    document_id: &str,
) -> ApiResult<UploadedDocument> {
    let app = load_owned(state, session, None).await?;
    let doc_uuid = parse_uuid_or_not_found(document_id)?;
    let found = repo::find_document(&state.pool, doc_uuid)
        .await?
        .filter(|f| f.application_id == Uuid::parse_str(&app.id).unwrap())
        .ok_or_else(|| ApiError::not_found("Document not found."))?;

    let mut tx = state.pool.begin().await?;
    let updated = repo::mark_document_uploaded(&mut tx, doc_uuid).await?;
    write_audit(
        &mut tx,
        AuditEntry {
            actor_id: Some(session.user_id),
            actor_role: Some(session.role_str()),
            action: "onboarding.document_retried",
            entity_type: "onboarding_document",
            entity_id: document_id.to_string(),
            before: Some(serde_json::to_value(&found.document)?),
            after: Some(serde_json::to_value(&updated)?),
        },
    )
    .await?;
    tx.commit().await?;
    Ok(updated)
}

pub async fn remove_document(
    state: &AppState,
    session: &Session,
    document_id: &str,
) -> ApiResult<()> {
    let app = load_owned(state, session, None).await?;
    let doc_uuid = parse_uuid_or_not_found(document_id)?;
    let found = repo::find_document(&state.pool, doc_uuid)
        .await?
        .filter(|f| f.application_id == Uuid::parse_str(&app.id).unwrap())
        .ok_or_else(|| ApiError::not_found("Document not found."))?;

    let mut tx = state.pool.begin().await?;
    repo::delete_document(&mut tx, doc_uuid).await?;
    write_audit(
        &mut tx,
        AuditEntry {
            actor_id: Some(session.user_id),
            actor_role: Some(session.role_str()),
            action: "onboarding.document_removed",
            entity_type: "onboarding_document",
            entity_id: document_id.to_string(),
            before: Some(serde_json::to_value(&found.document)?),
            after: None,
        },
    )
    .await?;
    tx.commit().await?;

    if let Some(key) = found.storage_key {
        let _ = storage::delete(&state.config.storage_dir, &key).await;
    }
    Ok(())
}

/// Slice: moves through submitted -> in_review (committed), runs the KYB
/// provider, then commits approved (+ approvedSummary) or rejected
/// (+ rejectionReasons). Resolves only at a terminal status — the frontend
/// does not poll.
pub async fn submit(
    state: &AppState,
    session: &Session,
    application_id: Option<String>,
) -> ApiResult<OnboardingApplication> {
    let app = load_owned(state, session, application_id.as_deref()).await?;
    let Some(business) = app.business.clone() else {
        return Err(ApiError::validation(
            "Add your business details before submitting.",
        ));
    };
    if app.status == "submitted" || app.status == "in_review" {
        return Err(ApiError::conflict(
            "This application is already being reviewed.",
        ));
    }
    let app_uuid = Uuid::parse_str(&app.id).unwrap();

    {
        let mut tx = state.pool.begin().await?;
        repo::patch_status(
            &mut tx,
            app_uuid,
            StatusPatch {
                status: "submitted".into(),
                mark_submitted: true,
                ..Default::default()
            },
        )
        .await?;
        audit_transition(&mut tx, session, &app.id, &app.status, "submitted").await?;
        repo::patch_status(
            &mut tx,
            app_uuid,
            StatusPatch {
                status: "in_review".into(),
                ..Default::default()
            },
        )
        .await?;
        audit_transition(&mut tx, session, &app.id, "submitted", "in_review").await?;
        tx.commit().await?;
    }

    let outcome = kyb::run_checks(&app, state.config.kyb_check_delay_ms).await;

    {
        let mut tx = state.pool.begin().await?;
        repo::replace_kyb_checks(&mut tx, app_uuid, &outcome.checks).await?;

        if outcome.approved {
            let summary = ApprovedAccountSummary {
                account_id: generate_account_id(&business.legal_name),
                risk_rating_label: "Medium-Low".into(),
                segment: business.industry.segment_label().into(),
                corridor: "NGN → USD / EUR".into(),
                monthly_limit: Money::new(100_000_00, CurrencyCode::Usd),
            };
            repo::patch_status(
                &mut tx,
                app_uuid,
                StatusPatch {
                    status: "approved".into(),
                    mark_reviewed: true,
                    approved_summary: Some(summary),
                    ..Default::default()
                },
            )
            .await?;
            audit_transition(&mut tx, session, &app.id, "in_review", "approved").await?;
        } else {
            repo::patch_status(
                &mut tx,
                app_uuid,
                StatusPatch {
                    status: "rejected".into(),
                    mark_reviewed: true,
                    rejection_reasons: Some(outcome.rejection_reasons),
                    ..Default::default()
                },
            )
            .await?;
            audit_transition(&mut tx, session, &app.id, "in_review", "rejected").await?;
        }
        tx.commit().await?;
    }

    load_owned(state, session, None).await
}

fn parse_uuid_or_not_found(value: &str) -> ApiResult<Uuid> {
    Uuid::parse_str(value).map_err(|_| ApiError::not_found("Document not found."))
}

async fn audit_transition(
    tx: &mut sqlx::PgConnection,
    session: &Session,
    app_id: &str,
    from: &str,
    to: &str,
) -> ApiResult<()> {
    write_audit(
        tx,
        AuditEntry {
            actor_id: Some(session.user_id),
            actor_role: Some(session.role_str()),
            action: "onboarding.state_change",
            entity_type: "onboarding_application",
            entity_id: app_id.to_string(),
            before: Some(json!({ "status": from })),
            after: Some(json!({ "status": to })),
        },
    )
    .await?;
    Ok(())
}
